test/nestedJoin.integration.test.ts

```diff
@@ -0,0 +1,528 @@
+                        {
+                            address:
+                                '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', // USDC
+                            decimals: 6,
+                            index: 2,
+                        },
+                        {
+                            address:
+                                '0xdac17f958d2ee523a2206206994597c13d831ec7', // USDT
+                            decimals: 6,
+                            index: 3,
+                        },
+                    ],
+                },
+            ],
+            mainTokens: [
+                {
+                    address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2', // WETH
+                    decimals: 18,
+                },
+                {
+                    address: '0x6b175474e89094c44da98b954eedeac495271d0f', // DAI
+                    decimals: 18,
+                },
+                {
+                    address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48', // USDC
+                    decimals: 6,
+                },
+                {
+                    address: '0xdac17f958d2ee523a2206206994597c13d831ec7', // USDT
+                    decimals: 6,
+                },
+            ],
+        };
+    }
+}
+
+export const grantRoles = async (
+    client: Client & TestActions & WalletActions,
+) => {
+    const balancerDaoAddress = '0x10A19e7eE7d7F8a52822f6817de8ea18204F2e4f';
+    const authorizerAddress = '0xA331D84eC860Bf466b4CdCcFb4aC09a1B43F3aE6';
+
+    // Check for available roles on balancer-deployments repo:
+    // https://github.com/balancer/balancer-deployments/blob/master/action-ids/mainnet/action-ids.json
+    const exitRole =
+        '0xc149e88b59429ded7f601ab52ecd62331cac006ae07c16543439ed138dcb8d34';
+    const joinRole =
+        '0x78ad1b68d148c070372f8643c4648efbb63c6a8a338f3c24714868e791367653';
+    const swapRole =
+        '0x7b8a1d293670124924a0f532213753b89db10bde737249d4540e9a03657d1aff';
+    const batchSwapRole =
+        '0x1282ab709b2b70070f829c46bc36f76b32ad4989fecb2fcb09a1b3ce00bbfc30';
+    const setRelayerApprovalRole =
+        '0x0014a06d322ff07fcc02b12f93eb77bb76e28cdee4fc0670b9dec98d24bbfec8';
+
+    await client.impersonateAccount({
+        address: balancerDaoAddress,
+    });
+    const roles: Address[] = [
+        exitRole,
+        joinRole,
+        swapRole,
+        batchSwapRole,
+        setRelayerApprovalRole,
+    ];
+    const chainId = await client.getChainId();
+    for (const role of roles) {
+        await client.writeContract({
+            account: balancerDaoAddress,
+            address: authorizerAddress,
+            chain: client.chain,
+            abi: authorizerAbi,
+            functionName: 'grantRole',
+            args: [role, BALANCER_RELAYER[chainId]],
+        });
+    }
+    await client.stopImpersonatingAccount({
+        address: balancerDaoAddress,
+    });
+};
+
+export const approveRelayer = async (
```

Comment:
Resued across the tests so could be moved to a shared helper.

Reply:
Will we wait for the new relayer to be deployed before merging this PR? I didn't pay much attention to organizing these helper methods because I was expecting they would eventually be removed.
I'll refactor just in case 👍

Reply:
I think we should wait otherwise the functionality won't work.